colossalai/communication/profiler.py

```diff
@@ -0,0 +1,270 @@
+                  "{:.1f}% of total communication time".format(event.self_cuda_time / self.total_cuda_time * 100.0),
+                  "self communication volme: {}".format(_format_memory(event.self_comm_vol)),
+                  "average bandwith: {}".format(_format_bandwith(event.self_comm_vol, event.self_cuda_time)),
+                  "number of calls: {}".format(event.self_count),
+                  "--------------------", sep='\n')
+
+    @property
+    def has_aync_op(self):
+        return self.pending_op is not None
+
+    def activate_profiler(self, kn: str, vol: float):
+        self.pending_metadata = (kn, _get_code_location(self.depth), vol)
+        self.profiler = profile(enabled=True, use_cuda=True,
+                                use_cpu=True, use_kineto=True)
+        self.profiler.__enter__()
+
+    def close_profiler(self, group=None):
+        assert self.profiler is not None, "There is no running dist op"
+        kernel_name, code_location, vol = self.pending_metadata
+        self.profiler.__exit__(None, None, None)
+
+        if self.profiler.enabled:
+            assert_flag = 0
+            current_comm_event = None
+            events = self.profiler.function_events
+            for event in events:
+                if kernel_name in event.name:
+                    assert assert_flag == 0, "Multiple dist ops has been called "
+                    current_comm_event = CommEvent(1, vol, event.self_cuda_time_total)
+                    assert_flag += 1
+
+            assert current_comm_event is not None, "dist op has not been found"
+
+            buffer = torch.tensor([current_comm_event.self_cuda_time],
+                                  device=get_current_device())
+            torch_all_reduce(buffer, op=ReduceOp.MIN, group=group)
+            current_comm_event.self_cuda_time = buffer.item()
+
+            self.total_count += current_comm_event.self_count
+            self.total_comm_vol += current_comm_event.self_comm_vol
+            self.total_cuda_time += current_comm_event.self_cuda_time
+            if code_location in self.ops_record:
+                self.ops_record[code_location].add(current_comm_event)
+            else:
+                self.ops_record[code_location] = current_comm_event
+
+        self.profiler = None
+        self.pending_op = None
+        self.pending_metadata = None
+
+    def wait_async_op(self):
+        if self.pending_op is not None:
+            op = self.pending_op
+            op.wait()
+            self.close_profiler()
+
+
+class CommHandler(object):
+
+    def __init__(self):
+        super().__init__()
+        self.prof = CCPROF
+
+    def wait(self):
+        self.prof.wait_async_op()
+
+
+CCPROF = CommProf()
+torch_all_reduce = dist.all_reduce
+torch_all_gather = dist.all_gather
+torch_reduce_scatter = dist.reduce_scatter
+torch_broadcast = dist.broadcast
+torch_reduce = dist.reduce
+
+
+def prof_init(depth: int = 0):

```

Comment:
Do write the method and class name properly for clarity, e.g. `initialize_profiling`. This applies to other naming in this PR as well, i.e. `CommProf`, `CCPROF`, `CommEvent`.

Reply:
Usually, COMM = COMMUNICATION. I have added comments to point out it.

Reply:
If this CommXXX is not exposed to the user, I can accept this.

Reply:
Users do not need to know anything about CommXXX, all they need to add is initialize_profiling and COL_COMM_PROF.show().

Reply:
Can we wrap `COL_COMM_PROF.show` in a function like `show_profile_results`, we should avoid letting users access the global constant.

Reply:
@1SAA

Reply:
Done